pkg/ccl/crosscluster/logical/logical_replication_job.go

```diff
@@ -172,46 +173,221 @@ func (r *logicalReplicationResumer) ingest(
 		req.TableIDs = append(req.TableIDs, pair.SrcDescriptorID)
 	}
 
-	plan, err := client.PlanLogicalReplication(ctx, req)
+	planner := makeLogicalReplicationPlanner(
+		ctx,
+		req,
+		jobExecCtx,
+		client,
+		progress,
+		payload,
+		jobID,
+		replicatedTimeAtStart)
+
+	initialPlan, initialPlanCtx, frontier, err := planner.generatePlanWithFrontier(ctx, distSQLPlanner)
 	if err != nil {
 		return err
 	}
 
+	replanOracle := sql.ReplanOnCustomFunc(
+		measurePlanChange,
+		func() float64 {
+			return crosscluster.LogicalReplanThreshold.Get(jobExecCtx.ExecCfg().SV())
+		},
+	)
+
+	replanner, stopReplanner := sql.PhysicalPlanChangeChecker(ctx,
+		initialPlan,
+		planner.generatePlan,
+		jobExecCtx,
+		replanOracle,
+		func() time.Duration { return crosscluster.LogicalReplanFrequency.Get(jobExecCtx.ExecCfg().SV()) },
+	)
+	metrics := execCfg.JobRegistry.MetricsStruct().JobSpecificMetrics[jobspb.TypeLogicalReplication].(*Metrics)
+
+	execPlan := func(ctx context.Context) error {
+		heartbeatSender := streamclient.NewHeartbeatSender(ctx, client, streampb.StreamID(streamID),
+			func() time.Duration {
+				return heartbeatFrequency.Get(&execCfg.Settings.SV)
+			})
+		defer func() {
+			_ = heartbeatSender.Stop()
+			stopReplanner()
+		}()
+
+		metaFn := func(_ context.Context, meta *execinfrapb.ProducerMetadata) error {
+			log.Warningf(ctx, "received unexpected producer meta: %v", meta)
+			return nil
+		}
+		rh := rowHandler{
+			replicatedTimeAtStart: replicatedTimeAtStart,
+			frontier:              frontier,
+			metrics:               metrics,
+			settings:              &execCfg.Settings.SV,
+			job:                   r.job,
+			frontierUpdates:       heartbeatSender.FrontierUpdates,
+		}
+		rowResultWriter := sql.NewCallbackResultWriter(rh.handleRow)
+		distSQLReceiver := sql.MakeDistSQLReceiver(
+			ctx,
+			sql.NewMetadataCallbackWriter(rowResultWriter, metaFn),
+			tree.Rows,
+			execCfg.RangeDescriptorCache,
+			nil, /* txn */
+			nil, /* clockUpdater */
+			evalCtx.Tracing,
+		)
+		defer distSQLReceiver.Release()
+		// Copy the evalCtx, as dsp.Run() might change it.
+		evalCtxCopy := *evalCtx
+		distSQLPlanner.Run(
+			ctx,
+			initialPlanCtx,
+			nil, /* txn */
+			initialPlan,
+			distSQLReceiver,
+			&evalCtxCopy,
+			nil, /* finishedSetupFn */
+		)
+
+		return rowResultWriter.Err()
+	}
+	err = ctxgroup.GoAndWait(ctx, execPlan, replanner)
+	if errors.Is(err, sql.ErrPlanChanged) {
+		metrics.ReplanCount.Inc(1)
+	}
+	return err
+}
+
+// Copying the PCR implementation for now. Will finetune based on testing.
```

Comment:
what other changes were you planning here that would make LDR's replanning policy different to PCRs here? If none, i suggest refactoring the PCR code to become more extensible, allowing PCR and LDR to use common code.

Reply:
moved measurePlanChange into distsql_plan_bulk.go and separated out getNodes. Thoughts?